cpp/oneapi/dal/detail/memory_impl_host.hpp

```diff
@@ -24,34 +24,32 @@
 namespace oneapi::dal::detail {
 namespace v1 {
 
+ONEDAL_EXPORT void* malloc_impl_host(const default_host_policy&, std::int64_t size);
+ONEDAL_EXPORT void free_impl_host(const default_host_policy&, void* pointer);
+ONEDAL_EXPORT void fill_impl_host(const default_host_policy&,
+                                  void* dest, std::int64_t size,
+                                  const void* pattern, std::int64_t pattern_size);
+
 template <typename T>
-inline T* malloc(const default_host_policy&, std::int64_t count) {
-    return new T[count];
+T* malloc(const default_host_policy& policy, std::int64_t count) {
+    const std::int64_t bytes_count = sizeof(T) * count;
+    ONEDAL_ASSERT(bytes_count > count);
+    return static_cast<T*>(malloc_impl_host(policy, bytes_count));
 }
 
 template <typename T>
-inline void free(const default_host_policy&, T* pointer) {
-    delete[] pointer;
+void free(const default_host_policy& policy, T* pointer) {
+    free_impl_host(policy, pointer);
 }
 
-inline void memset(const default_host_policy&, void* dest, std::int32_t value, std::int64_t size) {
-    // TODO: is not safe since std::memset accepts size as size_t
-    // TODO: can be optimized in future
-    std::memset(dest, value, size);
-}
-
-inline void memcpy(const default_host_policy&, void* dest, const void* src, std::int64_t size) {
-    // TODO: is not safe since std::memset accepts size as size_t
-    // TODO: can be optimized in future
-    std::memcpy(dest, src, size);
-}
+ONEDAL_EXPORT void memset(const default_host_policy&, void* dest, std::int32_t value, std::int64_t size);
+ONEDAL_EXPORT void memcpy(const default_host_policy&, void* dest, const void* src, std::int64_t size);
 
 template <typename T>
-inline void fill(const default_host_policy&, T* dest, std::int64_t count, const T& value) {
-    // TODO: can be optimized in future
-    for (std::int64_t i = 0; i < count; i++) {
-        dest[i] = value;
-    }
+void fill(const default_host_policy& policy, T* dest, std::int64_t count, const T& value) {
```

Comment:
```suggestion
inline void fill(const default_host_policy& policy, T* dest, std::int64_t count, const T& value) {
```

Reply:
`memset` can be replaced by `fill` in all callers, they duplicate logic. For optimization purpose, you can you `if constexpr` to check whether `T` is `std::int32_t`

Reply:
discussed offline not to apply this; these functions have a slightly different meaning